Schema: get all devenv dashboards passing validation

sdboyer opened the pull request:
This brings our schema to the point where all devenv dashboards having `schemaVersion` of at least 30 pass validation, meeting the goal set in #36844 🥳 🎉 🎉 🥳  

With all of those validating, the test mentioned in that issue is also enabled, which turns that from optimistic goal into CI-enforced policy: all new or updated devenv dashboards must continue to pass validation for all future PRs. it's the next step of tightening the ratchet towards full schematization!

There are some changes to the devenv files where things appeared to be slightly stale, as well as a couple changes to corresponding typescript types that seems like they must have been buggy.

Two shortcuts notable shortcuts i took here:
* I've specifically opted to not keep the cue files under `cue/` in sync with those under `packages/`. That's a crucial invariant we'll need to maintain eventually/soon, but because there are a number of things that need to change about how that mapping works + i don't have the knowledge to know where all those files under `components/` should necessarily go, i decided it was preferable to treat that as something we tackle with the next round of work, particularly with frontend's help.
* I've also opted not to hand-generate the `models.gen.ts` files for the new/updated `models.cue` files. Same basic reason - just trying to limit scope so that we can get more folks in on this. I could shift on this one and create/update them, though, if @ryantxu thinks it would make things notably better for frontenders

Fixes #36844

Commit:
Add transformations field

Commit:
Close FieldConfig struct; proper plugin validating

Commit:
s/graph/viz/ field in histogram dashboard

Commit:
Use ui.GraphFieldConfig in histogram model

Commit:
Add models for stat, gauge, barguage panel plugins

Also toss necessary shared types into cue/ui/gen.cue, with TODOs to move
them appropriately later.

Commit:
Add required license headers

Commit:
Heap of updates to cue UI components

Commit:
Fix barchart types and one old devenv input

Commit:
Use the GraphFieldConfig directly for timeseries

Commit:
Add models.cue for a few panel plugins

Barchart, state-timeline, and status-history

Commit:
Enable the test validating all devenv dashboards!!

ryantxu commented on cue/data/gen.cue:
```diff
@@ -113,6 +113,13 @@ Family: scuemata.#Family & {
                 } @cuetsy(targetType="interface")
 
+                // TODO docs
+                // FIXME this is extremely underspecfied; wasn't obvious which typescript types corresponded to it
```
See:
https://github.com/grafana/grafana/blob/main/public/app/features/dashboard/state/PanelModel.ts#L148

And:
https://github.com/grafana/grafana/blob/main/packages/grafana-data/src/types/transformations.ts#L32

I expect this is going to be:

```
#Transformation: [...{
    id: string
    options: {...}
}
```

Eventually the value of 'options' will have a whole registry system based on id -- there are a few more things that follow the same pattern.  Currently not extendable as plugins, but one day may be

sdboyer replied:
ahhhh, okie doke

sdboyer replied:
best to tackle in the next pass, i think

Commit:
Fix effects of not checking after making comments

ryantxu commented on packages/grafana-ui/src/options/models.gen.ts:
```diff
@@ -16,14 +16,14 @@ export interface VizTextDisplayOptions {
  */
 export interface OptionsWithLegend {
-  legend: VizLegendOptions;
+  legend?: VizLegendOptions;
```
We don't want to change this to optional

sdboyer replied:
Yeah, i was concerned about that. But validation fails here if it's required and has no default - there's devenv dashboards that are missing those values.

sdboyer replied:
and unfortunately these were some of the least helpful error messages - for some reason they DIDN'T come with the usual line numbers that CUE dumps out 😠 - something to def work on

ryantxu commented on packages/grafana-ui/src/options/models.gen.ts:
```diff
@@ -16,14 +16,14 @@ export interface VizTextDisplayOptions {
  */
 export interface OptionsWithLegend {
-  legend: VizLegendOptions;
+  legend?: VizLegendOptions;
```
```suggestion
  legend: VizLegendOptions;
```

sdboyer replied:
committed this so that we can see the error in CI

ryantxu commented on packages/grafana-ui/src/options/models.gen.ts:
```diff
@@ -16,14 +16,14 @@ export interface VizTextDisplayOptions {
  */
 export interface OptionsWithTooltip {
-  tooltip: VizTooltipOptions;
+  tooltip?: VizTooltipOptions;
```
```suggestion
  tooltip: VizTooltipOptions;
```

ryantxu commented on cue/ui/gen.cue:
```diff
@@ -93,3 +125,52 @@ VizLegendOptions: {
+} @cuetsy(targetType="interface")
+// TODO copy back to appropriate place
+OptionsWithLegend: {
+	legend?: VizLegendOptions
```
This is a good example where we should discuss what undefined/optional means.

I think we need to put in a default value for this and tooltip mode rather than have them optional

sdboyer replied:
I generally prefer the required-with-default approach over having it be optional, as well

Commit:
Update packages/grafana-ui/src/options/models.gen.ts

Co-authored-by: Ryan McKinley <ryantxu@gmail.com>

sdboyer commented:
Oh, wait, it didn't fail because your commit only updated the ts, not the cue.

Will need to circle back around to cleaning that (and the ts errors) up later

Commit:
Strip nulls (again)

Commit:
Add stripnulls script

Commit:
Realign and unalign cue with ts types

Commit:
Update devenv test to sniff for null errors

Best option we have right now for helping people to know they need to
strip nulls from devenv dashboards.

Commit:
Add speculative default for barchart stacking

Commit:
Fixup some dated devenv dashboards

timeline-modes needed to be regenerated with the appropriate tooltip
values included, per typing requirements, and timeline-demo needed to
have the `mode` field removed, as it is not intended to be persisted.

Commit:
Add necessary missing options for various panels

Commit:
Regenerate devenv dashboards

sdboyer commented:
There's now a bit more noise in this PR, but it's still all in pursuit of the original objective. Notes are mostly inline, but...

* I had neglected to disable openness on the `options` field. [Doing so](https://github.com/grafana/grafana/pull/37857/files#diff-920013f8bb5c45e5ac9590a88656b7f093813f1d63d60121baeaea806480d22cR224) revealed a number of other issues - missing declarations in schema that are fixed by other changes in the PR.

I added a script to strip nulls from devenv dashboards that have schema version 30 or 31. This is crucial now, since failure to strip nulls from devenv dashboards after this is merged will result in some rather confusing errors. In an attempt to mitigate that, i also added a heuristic in the Go test that runs all these validations on devenv dashboards to see if there appear to be null-related errors, so that simply reading test output can give an eagle-eyed observer a clue.

Null stripping should also probably be run as part of husky/lint-staged in order to make it properly part of local dev flow, but using the script to do so is iffy because of the reliance on GNU sed, and the need to install it on OSX - it's unacceptably bad to have an obscure `sed` error cause a `pre-commit` hook to fail because people have to `brew install gsed`. The proper approach there is probably to write a one-off JS function that `lint-staged` can call directly. But that requires refactoring to stop using `package.json` directives for lint-staged, and that's outside the scope of what i want to deal with here.

sdboyer commented on cue/data/gen.cue:
```diff
@@ -209,8 +218,10 @@ Family: scuemata.#Family & {
+                    // The allowable options are specified by the panel plugin's
+                    // schema.
+                    // FIXME same conundrum as with the closed validation for fieldConfig.
+                    options: {}
```
In the first pass at this PR, i neglected to disable openness on this `options` field. Doing so revealed a number of other issues - missing declarations in schema that are fixed by other changes in the recent commits.

sdboyer commented on cue/ui/gen.cue:
```diff
@@ -93,3 +125,54 @@ VizLegendOptions: {
+} @cuetsy(targetType="interface")
+// TODO copy back to appropriate place
+OptionsWithLegend: {
+	// FIXME this field is non-optional in the corresponding TS type
```
This FIXME seems the best thing to do for now, @ryantxu . My guess is that, until we've actually got cuetsy doing the conversions to TS (including those defaults), the conversation may be a bit too abstract.

Fortunately, that should be Coming Soon™ once this is in :)

sdboyer commented on pkg/schema/load/load_test.go:
```diff
@@ -96,7 +90,12 @@ func TestDevenvDashboardValidity(t *testing.T) {
+						errstr := errors.Details(err, nil)
+						t.Log(errstr)
+						if strings.Contains(errstr, "null") {
+							t.Log("validation failure appears to involve nulls - see if scripts/stripnulls.sh has any effect?")
```
This is the hail-mary pass that will give us a hint about null-related failures.

ryantxu replied:
progress not perfection

sdboyer commented:
This should now be fully ready for review - passing tests, no changes to TS files.

sdboyer commented on devenv/dev-dashboards/panel-graph/graph-ng-nulls.json:
```diff
@@ -8,6 +8,12 @@
         "hide": true,
```
All changes in this file result simply from re-exporting from the UI (and stripping nulls)

sdboyer commented on devenv/dev-dashboards/panel-timeline/timeline-demo.json:
```diff
@@ -8,6 +8,12 @@
         "hide": true,
```
All changes in this file result simply from re-exporting from the UI (and stripping nulls)

sdboyer commented on devenv/dev-dashboards/panel-timeline/timeline-demo.json:
```diff
@@ -168,9 +177,17 @@
+          "placement": "bottom"
+        },
+        "mergeValues": true,
         "mode": "changes",
```
Note that the `mode` here _is_ appearing sometimes, despite it not being intended to be part of the persisted data model. Something to fix in the frontend.

ryantxu commented on public/app/plugins/panel/timeseries/models.cue:
```diff
@@ -23,18 +23,12 @@ Family: {
             {
                 PanelOptions: {
+                    // FIXME idk where this is coming from but various devenv dashes have it
+                    graph?: {...}
```
looking though the dashboards, I think we can just search/replace:
```
"options": {
        "graph": {},
        "legend": {
```
with
```
"options": {
        "legend": {
```

if there are still issues, then lets figure that out, but looks like everything is empty object

sdboyer replied:
i should've been clearer in the comment - it's not just that devenv dashboards have it. They also still recreate it when re-exported from the frontend. So removing these now would create another bit of nasty one-off maintenance.

IMO better to change it in a follow-up when we fix the underlying behavior that's causing the field to still be generated.

ryantxu replied:
Hymm -- I just tried removeing all `graph` options -- including:
![image](https://user-images.githubusercontent.com/705951/129666632-49e57a29-469e-44d3-8cab-f8c069cdb239.png)

and I don't see it getting re-generated.  The string "realTimeUpdates" does not exist in our code :thinking: 

But I am totally on board merging this and then adding another issue to clean up more.
